crypto/asn1/d2i_pr.c

```diff
@@ -78,13 +78,50 @@ EVP_PKEY *d2i_PrivateKey(int type, EVP_PKEY **a, const unsigned char **pp,
  * type
  */
 
+static EVP_PKEY *key_as_pkcs8(const unsigned char **pp, long length, int *carry_on)
+{
+    const unsigned char *p = *pp;
+    PKCS8_PRIV_KEY_INFO *p8 = d2i_PKCS8_PRIV_KEY_INFO(NULL, &p, length);
+    EVP_PKEY *ret;
+
+    if (p8 == NULL)
+        return NULL;
+
+    ret = EVP_PKCS82PKEY(p8);
+    if (ret == NULL)
+        *carry_on = 0;
+
+    PKCS8_PRIV_KEY_INFO_free(p8);
+
+    if (ret != NULL)
+        *pp = p;
+
+    return ret;
+}
+
 EVP_PKEY *d2i_AutoPrivateKey(EVP_PKEY **a, const unsigned char **pp,
                              long length)
 {
     STACK_OF(ASN1_TYPE) *inkey;
     const unsigned char *p;
     int keytype;
+    EVP_PKEY *ret = NULL;
+    int carry_on = 1;
+
+    ret = key_as_pkcs8(pp, length, &carry_on);
```

Comment:
Should this be using ERR_set_mark()?

Reply:
Well it looks like you are trying this one - the if that fails then you try something else - For the case where you try something else the errors probably need to be cleared.

Reply:
I understand your point, but still not sure if
- we can get errors from my function
- we need to hide them: if we don't hide them, it provides more information about the formats that were tried.

Reply:
@beldmit - IMO the correct code structure would be:
1. ERR_set_mark();
2. key_as_pkcs8()
3. if OK : ERR_clear_last_mark(); return
4. d2i_PrivateKey()
5. if OK: ERR_pop_to_mark(); return
6. else: ERR_clear_last_mark(); return

Reply:
Fixed, thanks!